File: Go/Common/Sort/QuickSort/QuickSort.go
package main

import (
	"math/rand"
	"time"
	"fmt"
)

func randArr(len, maxNum int) []int {
	tempArr := []int{}
	rand.Seed(time.Now().Unix())
	for i := 0; i < len; i++ {
		x := rand.Intn(maxNum)
		tempArr = append(tempArr, x)
	}
	return tempArr
}

func quickSort(arr []int, left, right int) {
	if (left < right) {
		key := partition(arr, left, right)
		quickSort(arr, left, key-1)
		quickSort(arr, key+1, right)
	}
}

func partition(arr []int, left, right int) int {
	pivot := arr[right]
	j := left - 1
	for i := left; i <= right; i++ {
		if (arr[i] < pivot) {
			j++;
			arr[j], arr[i] = arr[i], arr[j]
		}
	}

	arr[j+1], arr[right] = arr[right], arr[j+1]
	return j + 1

}
func main() {
	arr := randArr(10, 50)
	for _, val := range arr {
		fmt.Printf("%d ", val)
	}
	fmt.Println()
	quickSort(arr, 0, len(arr)- 1)
	for _, val := range arr {
		fmt.Printf("%d ", val)
	}
	fmt.Println()
}

File: CPP/Common/Sort/Quick Sort/QuickSort.cpp
//
// Created by Insomnia on 2018/8/14.
// MIT License
//
#include <iostream>
#include <vector>

using namespace std;


void out(vector<int> &vec) {

    for (vector<int>::const_iterator it = vec.begin(); it != vec.end(); it++) {
        cout << (*it) << " ";
    }
    cout << endl;
}

vector<int> randomVec(int len, int maxNum) {
    vector<int> tempVec;
    srand((unsigned) time(NULL));
    for (int i = 0; i < len; ++i) {
        int x = rand() % maxNum;
        tempVec.push_back(x);
    }
    return tempVec;
}

class Solution {
public:
    void quickSort(vector<int> &vec, int left, int right) {
        if (left < right) {
            int pivot = partition(vec, left, right);
            quickSort(vec, left, pivot - 1);
            quickSort(vec, pivot + 1, right);
        }

    }

private:
    int partition(vector<int> &vec, int left, int right) {
        int pivot = vec[right];
        int j = left - 1;
        for (int i = left; i <= right; ++i) {
            if (vec[i] < pivot) {
                j++;
                swap(vec[j], vec[i]);
            }

        }

        swap(vec[j + 1], vec[right]);

        return j + 1;
    }
};

class Solution2 {
public:
    void quickSortWhile(vector<int> &vec, int left, int right) {
        int i = left;
        int j = right;
        int pivot = vec[left];

        while (i < j) {
            while (i < j && vec[j] >= pivot) {
                j--;
            }
            if (i < j) {
                swap(vec[i], vec[j]);
            }

            while (i < j && vec[i] <= pivot) {
                i++;
            }

            if (i < j) {
                swap(vec[i], vec[j]);
            }
        }

        if (i != left) {
            quickSortWhile(vec, left, i - 1);
        }
        if (j != right) {
            quickSortWhile(vec, j+1, right);
        }
    }
};

int main() {
//    vector<int> vec = randomVec(10, 60);
//    out(vec);
//    Solution solution;
//    solution.quickSort(vec, 0, vec.size() - 1);
//    out(vec);
    
    vector<int> vec = randomVec(10, 60);
    out(vec);
    Solution2 solution2;
    solution2.quickSortWhile(vec, 0, vec.size() - 1);
    out(vec);

    return 0;
}